Fix [dir=auto] invalidation with password fields

nt1m opened the pull request:
#### e8cd6950f18500186d21867ca5f3550f5bc4a429
<pre>
Fix [dir=auto] invalidation with password fields
<a href="https://bugs.webkit.org/show_bug.cgi?id=261941">https://bugs.webkit.org/show_bug.cgi?id=261941</a>
rdar://115887776

Reviewed by Aditya Keerthi.

Some text input types compute their directionality from the input value, some don&apos;t (password fields).

We did not invalidate when switching between these input types. This commit fixes that.

Inspired from <a href="https://chromium-review.googlesource.com/c/chromium/src/+/4851878">https://chromium-review.googlesource.com/c/chromium/src/+/4851878</a>

Spec: <a href="https://html.spec.whatwg.org/multipage/dom.html#auto-directionality">https://html.spec.whatwg.org/multipage/dom.html#auto-directionality</a>

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element.html:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-pseudo/dir-pseudo-on-input-element-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element-expected.txt:
* Source/WebCore/html/BaseTextInputType.cpp:
(WebCore::BaseTextInputType::dirAutoUsesValue const):
* Source/WebCore/html/BaseTextInputType.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::computeDirectionalityFromText const):
(WebCore::HTMLElement::updateTextDirectionalityAfterInputTypeChange):
(WebCore::HTMLElement::updateTextDirectionalityAfterTelephoneInputTypeChange): Deleted.
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::dirAutoUsesValue const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/html/HTMLTextFormControlElement.h:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::dirAutoUsesValue const):
* Source/WebCore/html/InputType.h:
* Source/WebCore/html/PasswordInputType.cpp:
(WebCore::PasswordInputType::dirAutoUsesValue const):
* Source/WebCore/html/PasswordInputType.h:

Canonical link: <a href="https://commits.webkit.org/268322@main">https://commits.webkit.org/268322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4482a9e27c265ff02c4409c33293fbc4674b358

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19350 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19768 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20367 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21239 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/18103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19585 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/23036 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/19893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19567 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/19614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/16823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/22109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/16800 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/17611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/22109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/17862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/17786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/22109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/18377 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/19893 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/17511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/17447 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/21871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2365 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/18207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element.html:
```diff
@@ -129,6 +129,30 @@
     }, `input element whose type attribute is in the ${type} state`);
 }
 
+test(() => {
```
Maybe this test case should go at the bottom of the file? The expected output seems a little strange with the unique assertion in the middle.

nt1m replied:
This was just imported from upstream, see https://chromium-review.googlesource.com/c/chromium/src/+/4851878

pxlcoder replied:
I see, shouldn't we be re-syncing separately?

nt1m replied:
We could, though it's more convenient in this case. Easier to integrate to the branch as well.

pxlcoder commented on Source/WebCore/html/HTMLElement.cpp:
```diff
@@ -834,8 +834,7 @@ std::optional<TextDirection> HTMLElement::directionalityIfDirIsAuto() const
     if (auto* textControl = dynamicDowncast<HTMLTextFormControlElement>(const_cast<HTMLElement*>(this))) {
-        auto* inputElement = dynamicDowncast<HTMLInputElement>(textControl);
-        if (!inputElement || (inputElement->isTextType() && !inputElement->isPasswordField())) {
+        if (textControl->dirAutoUsesValue()) {
```
This seems like a behavior change to include `<textarea>` – do we need to add a test for that?

nt1m replied:
It's not a behavior change, `!inputElement` covers `<textarea>`

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -536,13 +536,15 @@ void HTMLInputElement::updateType(const AtomString& typeAttributeValue)
-    if (oldType == InputType::Type::Telephone || m_inputType->type() == InputType::Type::Telephone)
-        updateTextDirectionalityAfterTelephoneInputTypeChange();
+    if (oldType == InputType::Type::Telephone || m_inputType->type() == InputType::Type::Telephone
+        || (hasDirectionAuto() && didDirAutoUseValue != m_inputType->dirAutoUsesValue()))
```
It might be nice to put this in a method that describes what this check is for. Up to you.

But at a minimum, we should probably include a link to the spec here?

nt1m replied:
Added a spec link

nt1m committed:
Fix [dir=auto] invalidation with password fields
https://bugs.webkit.org/show_bug.cgi?id=261941
rdar://115887776

Reviewed by Aditya Keerthi.

Some text input types compute their directionality from the input value, some don't (password fields).

We did not invalidate when switching between these input types. This commit fixes that.

Inspired from https://chromium-review.googlesource.com/c/chromium/src/+/4851878

Spec: https://html.spec.whatwg.org/multipage/dom.html#auto-directionality

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element.html:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-pseudo/dir-pseudo-on-input-element-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/selectors/dir-pseudo-on-input-element-expected.txt:
* Source/WebCore/html/BaseTextInputType.cpp:
(WebCore::BaseTextInputType::dirAutoUsesValue const):
* Source/WebCore/html/BaseTextInputType.h:
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::computeDirectionalityFromText const):
(WebCore::HTMLElement::updateTextDirectionalityAfterInputTypeChange):
(WebCore::HTMLElement::updateTextDirectionalityAfterTelephoneInputTypeChange): Deleted.
* Source/WebCore/html/HTMLElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::updateType):
(WebCore::HTMLInputElement::dirAutoUsesValue const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/html/HTMLTextFormControlElement.h:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::dirAutoUsesValue const):
* Source/WebCore/html/InputType.h:
* Source/WebCore/html/PasswordInputType.cpp:
(WebCore::PasswordInputType::dirAutoUsesValue const):
* Source/WebCore/html/PasswordInputType.h:

Canonical link: https://commits.webkit.org/268322@main
